### PR TITLE
fix(windows): consistent normalization in fs.find

### LINKED ISSUE #20594: path separator is not respected in vim.fs.find
### Neovim version (nvim -v)

0.9.0-dev-73-gcba05c541

### Vim (not Nvim) behaves the same?

NA

### Operating system/version

windows 11

### Terminal name/version

NA

### $TERM environment variable

NA

### Installation

NA

### How to reproduce the issue

```
:lua =vim.fs.find(".luacheckrc")
```

### Expected behavior

consistent slashes, either:
- `c:\\projects\\neovim\\.luacheckrc`
- `c:/projects/neovim/.luacheckrc`

### Actual behavior

`c:\\projects\\neovim/.luacheckrc`

### PR DESCRIPTION
**Pull request by kylo252**
fixes #20594

### How to test

run this command in the neovim repo root `:lua =vim.fs.find(".luacheckrc")`

before: `c:\\projects\\neovim/.luacheckrc`

with this PR: `c:/projects/neovim/.luacheckrc`

### CODE REVIEW

**Comment by kylo252** [comment]
@mfussenegger, what do you think of this change? 

I'm a bit split on normalizing slashes in the final result, but I guess it's a requirement?

**Comment by dundargoc** [comment]
Does this solve the same thing? https://github.com/neovim/neovim/pull/19988

**Comment by dundargoc** [comment]
OK, just wanted to check. Thanks.

File: runtime/lua/vim/fs.lua [hunk 1]
```diff
@@ -112,14 +112,15 @@ function M.find(names, opts)
   names = type(names) == 'string' and { names } or names
 
   local path = opts.path or vim.loop.cwd()
+  local path_sep = vim.loop.os_uname().version:match('Windows') and '\\' or '/'
```
Comment (ObserverOfTime) [inline runtime/lua/vim/fs.lua:115]: You can use [package.config](https://www.lua.org/manual/5.2/manual.html#pdf-package.config) to get the path separator:
```suggestion
  local path_sep = package.config:sub(1, 1)
```

Comment (glepnir) [reply runtime/lua/vim/fs.lua:115]: we have `vim.fn.windowsversion()` function.

Comment (Mike325) [reply runtime/lua/vim/fs.lua:115]: shouldn't this respect the windows' option `shellslash` ?

Comment (ObserverOfTime) [reply runtime/lua/vim/fs.lua:115]: > shouldn't this respect the windows' option `shellslash` ?

Indeed it should. That complicates things a bit, but this should work:

```lua
    local path_sep = '/'
    if vim.fn.windowsversion() ~= '' and not vim.o.shellslash then
       path_sep = '\\'
    end
```

Comment (glepnir) [reply runtime/lua/vim/fs.lua:115]: it's `#vim.fn.windowsversion() > 0 `

Comment (ObserverOfTime) [reply runtime/lua/vim/fs.lua:115]: > it's `#vim.fn.windowsversion() > 0`

I figured checking if `shellslash` exists would be better anyway.

Comment (glepnir) [reply runtime/lua/vim/fs.lua:115]: use the `windowsversion` to check is enough I think

Comment (kylo252) [reply runtime/lua/vim/fs.lua:115]: > shouldn't this respect the windows' option `shellslash` ?

I don't think libuv respect `shellslash` tho? I'm getting the same results for `vim.loop.cwd()` either way

```vim
:set shellslash
:lua =vim.loop.cwd() " -> 'C:\\projects`
:set noshellslash
:lua =vim.loop.cwd() " -> still 'C:\\projects`
```

so what's the point in checking that when `libuv` isn't affected in the first place?

File: runtime/lua/vim/fs.lua [hunk 2]
```diff
@@ -133,7 +134,7 @@ function M.find(names, opts)
         local t = {}
         for name, type in M.dir(p) do
           if names(name) and (not opts.type or opts.type == type) then
-            table.insert(t, p .. '/' .. name)
+            table.insert(t, p .. path_sep .. name)
```
[the remark this answers is not in the document]
Comment (kylo252) [reply runtime/lua/vim/fs.lua:137]: @justinmk, so which one of these should be the output?

> ### Expected behavior
> consistent slashes, either:
> 
> * `c:\\projects\\neovim\\.luacheckrc`
> * `c:/projects/neovim/.luacheckrc`
> 
> ### Actual behavior
> `c:\\projects\\neovim/.luacheckrc`

I wasn't worried about `shellslash` as per https://github.com/neovim/neovim/pull/20595#discussion_r993406846, this is just about producing a consistent behavior one way or another.

before: `c:\\projects\\neovim/.luacheckrc`

with this PR: `c:/projects/neovim/.luacheckrc`

we can get `c:\\projects\\neovim\\.luacheckrc`, if we skip the additional normalization I added 
```diff
diff --git a/runtime/lua/vim/fs.lua b/runtime/lua/vim/fs.lua
index 7bd635d8b692..ef41bbb49a0d 100644
--- a/runtime/lua/vim/fs.lua
+++ b/runtime/lua/vim/fs.lua
@@ -119,7 +120,7 @@ function M.find(names, opts)
 
   ---@private
   local function add(match)
-    matches[#matches + 1] = match
+    matches[#matches + 1] = M.normalize(match)
     if #matches == limit then
       return true
     end
```

[the remark this answers is not in the document]
Comment (justinmk) [reply runtime/lua/vim/fs.lua:137]: > For purposes of #20594 , forget about 'shellslash' and anything else in this PR. Let's focus on normalizing `\` slashes to `/` slashes.

**Comment by justinmk** [comment]
related: https://github.com/neovim/neovim/pull/20976

**Comment by dundargoc** [comment]
Is this ready to undraft?

**Comment by dundargoc** [comment]
Superseded by https://github.com/neovim/neovim/pull/22890

**Pull request closed without merging**